### PR TITLE
Support for changing team type

### LINKED ISSUE #2532: TeamTypes: enable upgrade path
### Description

Part of #2359 

We currently do not provide a way for changing team types. This is a requirement before we can launch the new team types in FF Cloud.

This will require an api endpoint to apply the change. This will need to check the change is valid. There are a number of scenarios (typically when 'downgrading') where the change will not be valid. For example:

 - A Team is exceeding a limit (users/devices/instances of a type)
 - A Team is using a feature that is not available in the target type. This could get quite hard to detect. For example, the limits on context/file storage - I don't think we have a way to query the current usage.

If billing is enabled, all of the stripe price/product ids need switching over. These include:
 - the Team billing item
 - the Device billing item
 - the billing items for each instance type

Note that if a team is in trial mode, the Team billing item may be using the trial price/product - this is an edge case that will need handling.

Once the API endpoint is in place, the UI needs implementing. We already have simple 'teamtype' choice when the user creates a new team. This likely needs some addition UX improvement and rework, but gives us a starting point.

### Epic/Story

2359

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate
```[tasklist]
### Tasks
- [x] Create API endpoint to modify team type
- [x] Create UX for modifying Team type
```

### PR DESCRIPTION
**Pull request by knolleary**
## Description

Closes #2532 

This adds the ability to change the Team Type.

```[tasklist]
### Tasks
- [x] Update api to support modifying team type - CE only
- [x] Add audit log entry type for changing team type
- [x] Add Change Type UI - CE only
- [x] Add EE/Billing logic around changing type
- [x] Update Change Type UI to give better billing change feedback
- [ ] Unit Test coverage of API changes
```

### CODE REVIEW

**Comment by knolleary** [comment]
I've added a bit of test coverage for the base model.

The existing test coverage of the `/api/v1/teams` end points is pretty bad. Will look to address that early next week.

Will also work on additional EE test coverage for the billing side.

File: forge/db/models/Team.js [hunk 1]
```diff
@@ -354,6 +354,87 @@ module.exports = {
...
                 checkInstanceStartAllowed: async function (instance) {
                     return true
+                },
+
+                /**
+                 * Checks whether the team type can be modified to the requested type.
+                 * The team must be within any limits of the target type.
+                 */
+                checkTeamTypeUpdateAllowed: async function (teamType) {
+                    // console.log(teamType.properties)
```
Comment (Pezmc) [inline forge/db/models/Team.js:364]: ✂️

File: forge/db/models/Team.js [hunk 2]
```diff
@@ -354,6 +354,87 @@ module.exports = {
...
+                            limit: targetDeviceLimit,
+                            count: currentDeviceCount
+                        })
+                    }
+
+                    const currentInstanceCountsByType = await this.instanceCountByType()
+                    // console.log(currentInstanceCountsByType)
+                    const targetInstanceLimits = {}
+                    for (const instanceType of Object.keys(currentInstanceCountsByType)) {
+                        // console.log('checking', instanceType, teamType.properties.instances[instanceType])
```
Comment (Pezmc) [inline forge/db/models/Team.js:399]: ✂️  Lots of this over the file, I won't comment on each one

File: forge/db/models/Team.js [hunk 3]
```diff
@@ -354,6 +354,87 @@ module.exports = {
...
+                    // }
+                },
+                /**
+                 * Update the team type.
+                 *
+                 * When running with EE, this function is replaced via ee/lib/billing/Team.js
+                 * to add additional checks related to billing
+                 */
+                updateTeamType: async function (teamType) {
+                    await this.setTeamType(teamType)
```
Comment (Pezmc) [inline forge/db/models/Team.js:435]: I'm slightly surprised that this method (updateTeamType) doesn't call `checkTeamTypeUpdateAllowed` and raise an exception if the change can't be made. Feels like it opens up the potential of team type being updated to an invalid state if `checkTeamTypeUpdateAllowed` is forgotten to be called first.

File: frontend/src/pages/team/Settings/General.vue [hunk 4]
```diff
@@ -12,7 +12,10 @@
         <FormRow v-model="input.teamType" type="uneditable">
             <template #default>Type</template>
             <template #description>
-                <div v-if="editing">You cannot currently change the type of team</div>
+                <template v-if="editing">
+                    Click <router-link :to="{name: 'TeamChangeType'}">here</router-link> to change the team type
+                </template>
+                <!-- <ff-button v-if="editing" size="small" :to="{name: 'TeamChangeType'}">Change</ff-button> -->
```
Comment (Pezmc) [inline frontend/src/pages/team/Settings/General.vue:18]: ✂️

**Review by Pezmc — APPROVED** [review]
Some nitpicks, but overall this looks good!

File: forge/db/models/Team.js (hunk 2 above)
Comment (knolleary) [reply forge/db/models/Team.js]: All tidied up

File: frontend/src/pages/team/Settings/General.vue (hunk 4 above)
Comment (knolleary) [inline frontend/src/pages/team/Settings/General.vue:18]:
```suggestion
```

**Comment by knolleary** [comment]
Still some EE test coverage to add - but moving forward to get this onto staging today to allow parallel testing

**Pull request merged**